src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp

```diff
@@ -0,0 +1,190 @@
+                     tmpl::next<tmpl::index_of<
+                         FaceVarsTagsList, tmpl::back<VolumeVarsTagsList>>>>>;
+  using front_face_vars_split = tmpl::split_at<
+      face_vars_excluding_extras_at_end,
+      tmpl::index_of<FaceVarsTagsList, tmpl::front<VolumeVarsTagsList>>>;
+  using volume_vars_face_subset_list = tmpl::back<front_face_vars_split>;
+  static_assert(
+      std::is_same_v<volume_vars_face_subset_list, VolumeVarsTagsList>,
+      "The VolumeVarsTagsList must be a subset of the FaceVarsTagsList.");
+  constexpr const size_t number_of_independent_components =
+      Variables<VolumeVarsTagsList>::number_of_independent_components;
+  using first_volume_tag = tmpl::front<VolumeVarsTagsList>;
+
+  const Mesh<Dim> uniform_gauss_mesh(volume_mesh.extents(0),
+                                     volume_mesh.basis(0),
+                                     Spectral::Quadrature::Gauss);
+  if (volume_mesh == uniform_gauss_mesh) {
+    const Matrix identity{};
+    auto interpolation_matrices = make_array<Dim>(std::cref(identity));
+    const auto& matrix = Spectral::boundary_interpolation_matrices(
+        volume_mesh.slice_through(direction.dimension()));
+    gsl::at(interpolation_matrices, direction.dimension()) =
+        direction.side() == Side::Upper ? matrix.second : matrix.first;
+
+    auto& first_face_field = get<first_volume_tag>(*face_fields);
+    auto& first_volume_field = get<first_volume_tag>(volume_fields);
+
+    DataVector face_view{
+        first_face_field[0].data(),
+        first_face_field[0].size() * number_of_independent_components};
+
+    apply_matrices(make_not_null(&face_view), interpolation_matrices,
+                   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+                   DataVector{const_cast<double*>(first_volume_field[0].data()),
+                              first_volume_field[0].size() *
+                                  number_of_independent_components},
+                   volume_mesh.extents());
+  } else {
+    ASSERT(Mesh<Dim>(volume_mesh.extents(0), volume_mesh.basis(0),
+                     Spectral::Quadrature::GaussLobatto) == volume_mesh,
+           "The mesh must be either a uniform Gauss or Gauss-Lobatto mesh, but "
+           "got "
+               << volume_mesh);
+    const size_t sliced_dim = direction.dimension();
+    const size_t fixed_index = direction.side() == Side::Upper
+                                   ? volume_mesh.extents(sliced_dim) - 1
+                                   : 0;
+
+    const size_t interface_grid_points =
+        volume_mesh.extents().slice_away(sliced_dim).product();
+    const size_t volume_grid_points = volume_mesh.number_of_grid_points();
+
+    const double* vars_data = volume_fields.data();
+    double* interface_vars_data = get<first_volume_tag>(*face_fields)[0].data();
+
+    // Run the SliceIterator as the outer-most loop since incrementing the slice
+    // iterator is surprisingly expensive.
```

Comment:
For my understanding:
- why not use the function `data_on_slice` here? this loop appears to be copy-pasted from there... please document why you are duplicating code if it is truly unavoidable.
- have you newly measured the slice iterator to be expensive, or are you documenting here the knowledge from your experiments ~2-3 years ago?

Reply:
- I'll add a comment: can't be done because `data_on_slice` assumes the face and volume tags are equal
- I have not, but the iterator hasn't changed since I optimised it and still concluded that it sucks :P